stackrox/stackrox
# fix(ui): Fix accessibility issue with non-unique banner landmark

### Description

### Problems reported by axe DevTools

> Ensures landmarks are unique

https://dequeuniversity.com/rules/axe/4.9/landmark-unique?application=AxeChrome

> Document should not have more than one banner landmark

https://dequeuniversity.com/rules/axe/4.9/landmark-no-duplicate-banner?application=AxeChrome

For both:

```html
<header class="pf-v5-c-page__header ignore-react-onclickoutside theme-dark">
```

Related node:

```html
<header class="pf-v5-c-modal-box__header">
```

### Analysis

1. Primary `header` element:

    * One suggestion that we can apply to make primary `header` element unique is `aria-label="primary"` attribute.

        https://developer.mozilla.org/en-US/blog/aria-accessibility-html-landmark-roles/

        > Here, we use the `<nav>` element element because it implicitly has the `navigation` role. Since a page could have multiple navigation elements, it is important to identify each of these landmarks uniquely. For this, we use the `aria-label` attribute and give the navigation landmark the name "Primary".

        ```html
        <nav aria-label="Primary">
        ```

        > Note When a screen reader describes the above navigation landmark, it will say "Primary navigation". This is why the word "navigation" is not included in the `aria-label` value.

    * For deprecated `PageHeader` element, `aria-label` prop is for sidebar toggle button.

    * For `Masthead` element, `aria-label` prop is for `header` element.

    Therefore, we can solve the first problem reported by axe DevTools.

2. Secondary `header` element of `Modal` or `Popup` elements:

    * Quick test of `header` prop of `Modal` element still renders `header` element.

    Therefore, unknown how to solve the second problem reported by axe DevTools.

### Solution

1. Rename and edit src/Containers/MainPage/Header/Header.tsx file.
    * Replace deprecated `PageHeader` with `Masthead` element and so on.
    * Add `className="pf-v5-u-flex-grow-1 pf-v5-u-justify-content-flex-end"` prop to approximate previous built-in style.
2. Edit src/Containers/MainPage/MastheadToolbar.tsx file.
    * Replace deprecated `PageHeaderTools` and so on with `Flex` and `FlexItem` elements.
    * Add `spaceItems={{ default: 'spaceItemsSm' }}` prop to approximate previous built-in style.
3. Edit src/Containers/MainPage/MainPage.tsx file.

### User-facing documentation


### Testing


#### Automated testing


#### How I validated my change

1. `yarn lint` in ui/apps/platform
2. `yarn build` in ui/apps/platform and then compute branch - master for the following
    * `wc build/static/css/*.css`
        main.js 44245 = 979712 - 935467
        total 44245 = 1969905 - 1925660
    * `ls -al build/static/css/*.css | wc`
        files 0 = 44 - 44
    * `wc build/static/js/*.js`
        main.js 2406 = 4617896 - 4615490
        total 2406 = 11701658 - 11699252
    * `ls -al build/static/js/*.js | wc`
        files 0 = 173 - 173
3. `yarn start` in ui/apps/platform

#### Manual testing

1. Visit /main/network-graph, and then click **Manage CIDR blocks** button.

    * Before changes, see both **landmark** accessibility issues.
        ![network-graph_with_both_issues](https://github.com/user-attachments/assets/29728cd7-3cbd-4400-8682-8f4cc65cc61c)

    * After changes, see only one **landmark** accessibility issues.
        ![network-graph_with_one_issue](https://github.com/user-attachments/assets/233a9a15-20e9-41b9-951d-81a6f169f720)

2. Visit /main/vulnerabilities/reports, click **Create report**, click **Select a collection** and then click **Create collection**.

    * Before changes, see both **landmark** accessibility issues.
        ![reports_with_both_issues](https://github.com/user-attachments/assets/4db7d9d5-7a78-4e0b-8a40-113c3f35c157)

    * After changes, see both **landmark** accessibility issues, because the form itself has non-unique landmarks.
        ![reports_with_both_issues_for_sections](https://github.com/user-attachments/assets/cd57c053-9d96-4583-9355-a976f4b958f5)

3. Visit /main/system-health, and then click **Show administrative usage** button.

    * Before changes, see both **landmark** accessibility issues.
        ![system-health_with_both_issues](https://github.com/user-attachments/assets/3c8d7cfa-93dd-4d2e-a869-fbbda4e7f4ac)

    * After changes, see only one **landmark** accessibility issues.
        ![system-health_with_one_issue](https://github.com/user-attachments/assets/f527cdc7-f5b8-4725-bf9e-a03fbcf9b202)

### Integration testing

* main/HelpMenu.test.js
* logout.test.js
* userinfo.test.js
